More evolution tests

slavapestov opened the pull request.

slavapestov committed:
Sema: More consistent @_fixed_layout diagnostics

slavapestov committed:
Evolution: Rename some tests in preparation for testing @_fixed_layout classes more

slavapestov committed:
Evolution: Add test changing a lazy property a computed property with a mutating getter

slavapestov committed:
Evolution: Test moving methods between class/struct and extension

slavapestov committed:
Evolution: Test adding a class deinit

slavapestov committed:
Evolution: Test that adding or removing weak/unowned is a resilient change

slavapestov committed:
Evolution: Test that adding willSet/didSet to a stored property works

slavapestov committed:
Evolution: Test adding a new convenience init to an open class

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on validation-test/Evolution/Inputs/class_change_lazy_to_computed.swift:
```diff
@@ -0,0 +1,33 @@
+
+#if BEFORE
```
These are backwards, no? BEFORE should be the lazy one. (I mean, we're testing in both directions, but the names should match the code.)

slavapestov replied:
Done

jrose-apple commented on validation-test/Evolution/test_class_add_deinit.swift:
```diff
@@ -0,0 +1,37 @@
+// RUN: %target-resilience-test
```
Do we ever test this optimized? Because that's really where I'd be concerned about this…and probably a bunch of the other ones too.

slavapestov replied:
That's a good idea. I'll have it run all test with and without -O.

jrose-apple commented on validation-test/Evolution/test_class_add_convenience_init.swift:
```diff
@@ -0,0 +1,20 @@
+}
+
+AddConvenienceInitTest.test("AddConvenienceInit") {
+  let t = Hobbit(z: 4)
```
Can you make sure it's callable too, in the AFTER mode?

slavapestov replied:
The test file itself is linked with both the before and after versions so it cannot call a symbol that's only available in the after library (unless we use `@_weakLinked`, I guess). The new convenience init is called by other init though. Did you have something else in mind?

jrose-apple replied:
I mean, we certainly don't want to break existing code, but we also want to make sure that the thing that was added is usable and backwards-deployable.

slavapestov committed:
Evolution: Build tests with optimizations if required

Make sure we honor the swift_test_mode parameter, by passing in
the right optimization flags. Also if we're passing in any flags,
pass in -wmo too.

slavapestov commented:
@swift-ci Please smoke test
